src/QMCDrivers/DMC/DMCBatched.cpp

```diff
@@ -417,10 +437,16 @@ void DMCBatched::runDMCStep(int crowd_id,
                             const StateForThread& sft,
                             DriverTimers& timers,
                             //                            DMCTimers& dmc_timers,
-                            std::vector<std::unique_ptr<ContextForSteps>>& context_for_steps,
-                            std::vector<std::unique_ptr<Crowd>>& crowds)
+                            UPtrVector<ContextForSteps>& context_for_steps,
+                            UPtrVector<Crowd>& crowds)
 {
-  Crowd& crowd  = *(crowds[crowd_id]);
+  Crowd& crowd                         = *(crowds[crowd_id]);
+  auto setThreadLocalRNGForHamiltonian = [](QMCHamiltonian& ham, ContextForSteps& step_context) {
+    ham.setRandomGenerator(&(step_context.get_random_gen()));
+  };
+  for (int iw = 0; iw < crowd.size(); ++iw)
+    setThreadLocalRNGForHamiltonian(crowd.get_walker_hamiltonians()[iw], *(context_for_steps[iw]));
```

Comment:
Can we move this added part into crowd as a member function? This part has been copied many times.

Reply:
`context_for_steps[iw]` or `context_for_steps[crowd_id]`?
`crowd.setRNGForHamiltonian(context_for_steps[crowd_id]->get_random_gen());`

Reply:
I really need to try this branch so just pushed the bug fix.